File: taller/IvanRC-taller-Java/src/main/java/modelsTaller/Taller.java
package modelsTaller;

import factoryPersonasTaller.FactoryPersonas;

public class Taller {

    public Taller(int numeroDeTrabajadores){
        this.numeroDeTrabajadores = numeroDeTrabajadores;
    }


    public final int JEFE_TALLER = 2500;
    public final int TRABAJADOR = 1200;
    public final int CHAPISTA = 1700;
    public final int ELECTRICISTA = 1800;
    public final int NAVAJA_SUIZA = 2000;

    public int numeroDeTrabajadores;
    public Persona[] arrayTaller = null;

    /**
     * función que sirve para rellenar el array de personas, con personas aleatorias
     */
    public void inicializarArray(){
        arrayTaller = new Persona[numeroDeTrabajadores];
        for(int i=0;i< arrayTaller.length;i++){
            arrayTaller[i] = FactoryPersonas.getInstance().createRandomPersona();
        }
    }

    /**
     * función que sirve para sacar por pantalla una lista de las personas de taller
     */
    public void representarListaPersonas() {
        for(int i=0;i<arrayTaller.length;i++){
            System.out.println(arrayTaller[i]);
        }
    }

    /**
     * función que sirve para hallar la nómina total en el taller
     * @return la nómina calculada
     */
    public int calcularNominaTotal(){
        var resultado = 0;
        for(int i=0;i<arrayTaller.length;i++){
            if(arrayTaller[i] instanceof Persona.JefeTaller){
                resultado += JEFE_TALLER;
            }else{
                if(arrayTaller[i] instanceof Persona.Trabajador && (arrayTaller[i] instanceof Persona.Electricista) == false && (arrayTaller[i] instanceof Persona.Chapista) == false && (arrayTaller[i] instanceof Persona.NavajaSuiza) == false){
                    resultado += TRABAJADOR;
                }else{
                    if(arrayTaller[i] instanceof Persona.Chapista){
                        resultado += CHAPISTA;
                    }else{
                        if(arrayTaller[i] instanceof Persona.Electricista){
                            resultado += ELECTRICISTA;
                        }else{
                            resultado += NAVAJA_SUIZA;
                        }
                    }
                }
            }
        }
        return resultado;
    }

    /**
     * función que calcula el número de jefes de taller que hay en el taller
     * @return el número de jefes de taller
     */
    public int calcularNumeroDeJefesTaller(){
        var contador = 0;
        for(int i=0;i<arrayTaller.length;i++){
            if(arrayTaller[i] instanceof Persona.JefeTaller) contador++;
        }
        return contador;
    }

    /**
     * función que calcula el número de trabajadores que hay en el taller
     * @return el número de trabajadores
     */
    public int calcularNumeroDeTrabajadores(){
        var contador = 0;
        for(int i=0;i<arrayTaller.length;i++){
            if(arrayTaller[i] instanceof Persona.Trabajador && (arrayTaller[i] instanceof Persona.Electricista) == false && (arrayTaller[i] instanceof Persona.Chapista) == false && (arrayTaller[i] instanceof Persona.NavajaSuiza) == false) contador++;
        }
        return contador;
    }

    /**
     * función que calcula el número de chapistas que hay en el taller
     * @return el número de chapistas
     */
    public int calcularNumeroDeChapistas(){
        var contador = 0;
        for(int i=0;i<arrayTaller.length;i++){
            if(arrayTaller[i] instanceof Persona.Chapista) contador++;
        }
        return contador;
    }

    /**
     * función que calcula el número de electricistas que hay en el taller
     * @return el número de electricistas
     */
    public int calcularNumeroDeElectricistas(){
        var contador = 0;
        for(int i=0;i<arrayTaller.length;i++){
            if(arrayTaller[i] instanceof Persona.Electricista) contador++;
        }
        return contador;
    }

    /**
     * función que calcula el número de navajas suizas que hay en el taller
     * @return el número de navajas suizas
     */
    public int calcularNumeroDeNavajasSuizas(){
        var contador = 0;
        for(int i=0;i<arrayTaller.length;i++){
            if(arrayTaller[i] instanceof Persona.NavajaSuiza) contador++;
        }
        return contador;
    }
}

File: taller/IvanRC-taller-Kotlin/src/main/kotlin/modelsTaller/Taller.kt
package modelsTaller

import factoryPersonasTaller.FactoryPersonas

class Taller(val numeroPersonas: Int) {
    val JEFE_TALLER = 2500
    val TRABAJADOR = 1200
    val CHAPISTA = 1700
    val ELECTRICISTA = 1800
    val NAVAJA_SUIZA = 2000

    val  arrayTaller: Array<Persona?> = Array(numeroPersonas){ FactoryPersonas.getInstance()!!.createRandomPersona()}

    /**
     * función que sirve para sacar por pantalla una lista de las personas de taller
     */
    fun representarListaPersonas() {
        for(i in arrayTaller){
            println(i)
        }
    }

    /**
     * función que sirve para hallar la nómina total en el taller
     * @return la nómina calculada
     */
    fun calcularNominaTotal(): Int{
        var resultado = 0
        for(i in arrayTaller){
            when(i){
                is Persona.JefeTaller -> resultado += JEFE_TALLER
                is Persona.Chapista -> resultado += CHAPISTA
                is Persona.Electricista -> resultado += ELECTRICISTA
                is Persona.NavajaSuiza -> resultado += NAVAJA_SUIZA
                is Persona.Trabajador -> resultado += TRABAJADOR
                else -> {}
            }
        }
        return resultado
    }

    /**
     * función que calcula el número de jefes de taller que hay en el taller
     * @return el número de jefes de taller
     */
    fun calcularNumeroDeJefesTaller(): Int{
        var contador = 0
        for(i in arrayTaller){
            if(i is Persona.JefeTaller) contador++
        }
        return contador
    }

    /**
     * función que calcula el número de trabajadores que hay en el taller
     * @return el número de trabajadores
     */
    fun calcularNumeroDeTrabajadores(): Int{
        var contador = 0
        for(i in arrayTaller){
            if(i is Persona.Trabajador && i !is Persona.Electricista && i !is Persona.Chapista && i !is Persona.NavajaSuiza) contador++
        }
        return contador
    }

    /**
     * función que calcula el número de chapistas que hay en el taller
     * @return el número de chapistas
     */
    fun calcularNumeroDeChapistas(): Int{
        var contador = 0
        for(i in arrayTaller){
            if(i is Persona.Chapista) contador++
        }
        return contador
    }

    /**
     * función que calcula el número de electricistas que hay en el taller
     * @return el número de electricistas
     */
    fun calcularNumeroDeElectricistas(): Int{
        var contador = 0
        for(i in arrayTaller){
            if(i is Persona.Electricista) contador++
        }
        return contador
    }

    /**
     * función que calcula el número de navajas suizas que hay en el taller
     * @return el número de navajas suizas
     */
    fun calcularNumeroDeNavajasSuizas(): Int{
        var contador = 0
        for(i in arrayTaller){
            if(i is Persona.NavajaSuiza) contador++
        }
        return contador
    }
}